src/saga_action_generic.rs

```diff
@@ -21,18 +21,6 @@ use std::sync::Arc;
  * bounds.
  */
 pub trait SagaType: Debug + 'static {
-    /**
```

Comment:
Saga Parameters now live inside the `Dag` and are output by Dag nodes. This allows the parent saga and each appended subsaga to have their own parameters. Unfortunately, the type-checking is only dynamic when looking up output nodes via `ActionContext::lookup`, which must deserialize a JSON value to the concrete type. However, this latter part was always the case. 

While removing some type safety, we add functionality, because now any action can be used in any saga node across all sagas. What we really need is to enable creating a single DAG made up of sagas/subsagas. If we can put some type safety back later, and lose the ability to share actions across saga types that may be a fair exchange. However, I'd like to offer an alternative suggestion that will give us the ability to exhaustively type-check all node outputs and lookups for any dynamically created DAG. 

The idea is that we create an attribute proc macro that annotates `ActionFunc` functions that perform `Action::do_it` and `Action::undo_it` methods. This macro will generate new functions with name appended with `_stub`, when the code is compiled in `test` configuration. The `Action` trait will also have two new members `do_it_stub` and `undo_it_stub` that get defined in `test` configuration. A macro (declarative is probably fine here), can be used to add these methods to any impl'd action by using the known stub names of the functions called in `do_it` and `undo_it`. At this point, we'd have the actions in the registry and their stub methods available. 

So what do the stub methods do? The only thing they do is to output a `Default` instance of the output type of the action, and also to call any `ActionContext::lookup` methods such that those default instances can be retrieved and deserialized. We then add functionality to the `SagaExecutor` such that when a given environment variable is defined (and we are in test configuration), the stub methods are run rather than the real methods. We then write a simple test that runs our "real" saga DAGs with the stub methods. We can probably wrap the `DagBuilder` in another macro and autogenerate the tests. These tests would execute any dynamically generated DAG and ensure that all lookup calls can find ancestor nodes with the correct output types. We thus have complete type-checking, although it's performed by a test rather than the rust compiler.

Reply:
Think I'm going to punt on this, and instead focus on writing property based test cases for the full sagas.